README.md
Lightweight UDP library for Dart.

## Usage

A simple usage example:

```dart
import 'package:udp/udp.dart';


main() async {
    // creates a UDP instance and binds it to the first available network
    // interface on port 65000.
    var sender = await UDP.bind(Endpoint.any(port: Port(65000)));
  
    // send a simple string to a broadcast endpoint on port 65001.
    var dataLength = await sender.send("Hello World!".codeUnits,
    Endpoint.broadcast(port: Port(65001)));
  
    stdout.write("$dataLength bytes sent.");
  
    // creates a new UDP instance and binds it to the local address and the port
    // 65002.
    var receiver = await UDP.bind(Endpoint.loopback(port: Port(65002)));
  
    // receiving\listening
    receiver.asStream(timeout: Duration(seconds: 20)).listen((datagram) {
      var str = String.fromCharCodes(datagram.data);
      stdout.write(str);
    });
  
    // close the UDP instances and their sockets.
    sender.close();
    receiver.close();
  
  
   // MULTICAST
    var multicastEndpoint =
        Endpoint.multicast(InternetAddress("239.1.2.3"), port: Port(54321));
  
    var receiver = await UDP.bind(multicastEndpoint);
  
    var sender = await UDP.bind(Endpoint.any());
  
    receiver.asStream().listen((datagram) {
      if (datagram != null) {
        var str = String.fromCharCodes(datagram?.data);
  
        stdout.write(str);
      }
    });
  
    await sender.send("Foo".codeUnits, multicastEndpoint);
  
    await Future.delayed(Duration(seconds:5));
  
    sender.close();
    receiver.close();
  
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/xenoken/udp/issues


CHANGELOG.md
## 5.0.3
- Documentation improvements.

## 5.0.2
- Fixed an exception occurring when binding UDP receivers to a multicast endpoint.
- Added additional documentation for Endpoints.
- Minor changes.

## 5.0.1
- Updated README.md.

## 5.0.0
- removed "UDP.listen" method.
- added "UDP.asStream" method.
- Minor changes.

## 4.0.1
- Fixed formatting

## 4.0.0
- Migration to null safety.
- Minor changes.

## 3.0.3
- Changes to README.md

## 3.0.2

- Package Maintenance.

## 3.0.1

- Minor improvements.

## 3.0.0

- Added: Multicast support !
- Change: the 'timeout' parameter of 'UDP.listen' is now optional. UDP instances can listen indefinitely now.
- Change: 'Endpoint.IsBroadcast' is now a readonly property.
- Change: in 'Endpoint.unicast', port is now optional. 
- Change: 'UDP.listen' returns a Future<bool>.

## 2.0.0

- Added: 'UDP.closed' property: readonly property that is True on closed UDP instances.
- Added: It is not possible to call 'UDP.listen' on a udp instance that was already listening.
- Added: 'UDP.send' returns -1 on closed instances.
- Change: UDPReceiveCallback renamed to DatagramCallback.

## 1.0.2

- Package Maintenance: description rewritten.


## 1.0.1

- Minor improvements.


## 1.0.0

- Initial version.


test/udp_test.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha 
 *  
 *  Redistribution and use in source and binary forms, with or without 
 *  modification, are permitted provided that the following conditions 
 *  are met: 
 *  
 *  1. Redistributions of source code must retain the above copyright 
 *  notice, this list of conditions and the following disclaimer. 
 *  
 *  2. Redistributions in binary form must reproduce the above copyright 
 *  notice, this list of conditions and the following disclaimer in the 
 *  documentation and/or other materials provided with the distribution. 
 *  
 *  3. Neither the name of the copyright holder nor the names of its 
 *  contributors may be used to endorse or promote products derived from 
 *  this software without specific prior written permission. 
 *  
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY 
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 *  POSSIBILITY OF SUCH DAMAGE.
 *  
 */

import 'dart:cli';
import 'dart:io';

import 'package:test/test.dart';
import 'package:udp/udp.dart';

void main() {
  group('udp', () {
    /*
    UNICAST
    */
    test('Unicast', () async {
      var result = '';

      UDP? receiver, sender;

      waitFor(Future.wait([
        UDP
            .bind(
                Endpoint.unicast(InternetAddress('127.0.0.1'), port: Port(42)))
            .then((udp) {
          receiver = udp;
          return udp.asStream(timeout: Duration(seconds: 5)).listen((dgram) {
            result = String.fromCharCodes(dgram!.data);
          });
        }),
        UDP
            .bind(
                Endpoint.unicast(InternetAddress('127.0.0.1'), port: Port(24)))
            .then((udp) {
          sender = udp;
          return udp.send('Foo'.codeUnits, Endpoint.broadcast(port: Port(42)));
        })
      ]));

      receiver?.close();
      sender?.close();

      expect(result, equals('Foo'));
    });

    /*
    LOOPBACK
    */
    test('Loopback', () async {
      var result = '';

      UDP? receiver, sender;

      waitFor(Future.wait([
        UDP.bind(Endpoint.loopback(port: Port(42))).then((udp) {
          receiver = udp;
          return udp.asStream(timeout: Duration(seconds: 5)).listen((dgram) {
            result = String.fromCharCodes(dgram!.data);
          });
        }),
        UDP.bind(Endpoint.loopback(port: Port(24))).then((udp) {
          sender = udp;
          return udp.send('Foo'.codeUnits, Endpoint.loopback(port: Port(42)));
        })
      ]));

      receiver?.close();
      sender?.close();

      expect(result, equals('Foo'));
    });

    /*
    BROADCAST
    */
    test('Broadcast', () async {
      var result = '';

      UDP? receiver, sender;
      waitFor(Future.wait([
        UDP.bind(Endpoint.any(port: Port(42))).then((udp) {
          receiver = udp;
          return udp.asStream(timeout: Duration(seconds: 5)).listen((dgram) {
            result = String.fromCharCodes(dgram!.data);
          });
        }),
        UDP.bind(Endpoint.any(port: Port(24))).then((udp) {
          sender = udp;
          return udp.send('Foo'.codeUnits, Endpoint.broadcast(port: Port(42)));
        })
      ]));

      receiver?.close();
      sender?.close();

      expect(result, equals('Foo'));
    });

    /*
    MULTICAST
    */
    test('Multicast', () async {
      var result = '';

      UDP? receiver, sender;

      var multicastEndpoint =
          Endpoint.multicast(InternetAddress('239.1.2.3'), port: Port(4540));

      waitFor(Future.wait([
        UDP.bind(multicastEndpoint).then((udp) {
          return udp.asStream(timeout: Duration(seconds: 10)).listen((dgram) {
            if (dgram == null) return;
            result = String.fromCharCodes(dgram.data);
          });
        }),
        UDP.bind(Endpoint.any()).then((udp) {
          sender = udp;
          return sender!.send('Foo'.codeUnits, multicastEndpoint);
        })
      ]));

      receiver?.close();
      sender?.close();

      expect(result, equals('Foo'));
    });

    /*
    Second listen is not possible.
    */
    test('Second Listen on the same instance should not be possible.',
        () async {
      var udp = await UDP.bind(Endpoint.loopback());

      var receiver = await UDP.bind(Endpoint.loopback());

      var value = 'original';

      receiver.asStream(timeout: Duration(seconds: 5)).listen((datagram) {
        print(String.fromCharCodes(datagram!.data));
      });

      // this listen request doesn't do anything.
      receiver.asStream(timeout: Duration(seconds: 5)).listen((datagram) {
        value = 'modified';
        print(String.fromCharCodes(datagram!.data));
      });

      await udp.send('Foo'.codeUnits, Endpoint.broadcast());

      receiver.close();

      udp.close();

      expect(value == 'original', isTrue);
    });

    /*
    A closed UDP instance can't be reused.
    */
    test('Using a closed UDP instance is not possible.', () async {
      var udp = await UDP.bind(Endpoint.loopback());

      var receiver = await UDP.bind(Endpoint.loopback());

      receiver.close();

      udp.close(); // trying to see what happens if a send or receive method is called on a closed udp instance.

      receiver.asStream(timeout: Duration(seconds: 5)).listen((datagram) {
        print(String.fromCharCodes(datagram!.data));
      });

      var dataLength = await udp.send('Foo'.codeUnits, Endpoint.broadcast());

      expect(dataLength == -1, isTrue);
    });

    /*
    UDP.Close() sets UDP.closed to TRUE
    */
    test('The property "closed" is True for closed udp instances.', () async {
      var udp = await UDP.bind(Endpoint.loopback());

      var receiver = await UDP.bind(Endpoint.loopback());

      receiver.close();

      udp.close();

      expect(receiver.closed && udp.closed, isTrue);
    });

    /*
    UDP listen can run forever if no timeout is set.
    */
    test('UDP listen can run forever if no timeout is set.', () async {
      var receiver = await UDP.bind(Endpoint.any());

      receiver.asStream(timeout: null).listen((datagram) {});

      await Future.delayed(Duration(seconds: 5));

      expect(receiver.closed, isFalse);

      receiver.close();
    });

    /*
    A UDP instance listening indefinitely can be stopped by close.
    */
    test(' A UDP instance listening indefinitely can be stopped by close.',
        () async {
      var receiver = await UDP.bind(Endpoint.any());

      receiver.asStream(timeout: Duration(seconds: 5)).listen((datagram) {});

      await Future.delayed(Duration(seconds: 10));

      receiver.close();

      expect(receiver.closed, isTrue);
    });
  });
}


lib/udp.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright
 *  notice, this list of conditions and the following disclaimer in the
 *  documentation and/or other materials provided with the distribution.
 *
 *  3. Neither the name of the copyright holder nor the names of its
 *  contributors may be used to endorse or promote products derived from
 *  this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *  
 */
library udp;

export 'src/udp_base.dart' show UDP;
export 'src/udp_endpoint.dart' show Endpoint;
export 'src/udp_port.dart' show Port;


lib/src/udp_base.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright
 *  notice, this list of conditions and the following disclaimer in the
 *  documentation and/or other materials provided with the distribution.
 *
 *  3. Neither the name of the copyright holder nor the names of its
 *  contributors may be used to endorse or promote products derived from
 *  this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 */

import 'dart:async';
import 'dart:io';
import 'dart:collection';

import 'udp_endpoint.dart';

typedef DatagramCallback = void Function(Datagram?);

/// [UDP] sends or receives UDP packets.
///
/// a [UDP] instance can send packets to or receive packets from [Endpoint]s.
class UDP {
  /// Returns True if this [UDP] instance is closed.
  bool get closed => _closed;

  /// The [Endpoint] this [UDP] instance is bound to.
  Endpoint get local => _localep;

  /// Reference to underlying [RawDatagramSocket].
  RawDatagramSocket? get socket => _socket;

  // Reference to the local endpoint
  final Endpoint _localep;

  // Reference to the internal socket
  RawDatagramSocket? _socket;

  // Reference to the socket broadcast stream
  Stream? _socketBroadcastStream;

  // Reference to the UDP instance broadcast stream
  Stream<Datagram?>? _udpBroadcastStream;

  // Reference to the internal stream controller
  StreamController? _streamController;

  // Stores the set of internal stream subscriptions
  final HashSet<StreamSubscription> _streamSubscriptions =
      HashSet<StreamSubscription>();

  // Is the UDP instance closed?
  bool _closed = false;

  // Internal ctor
  UDP._(this._localep);

  /// Creates a new [UDP] instance.
  ///
  /// The [UDP] instance is created by the OS and bound to a local [Endpoint].
  ///
  /// [ep] - the local endpoint.
  ///
  /// returns the [UDP] instance.
  static Future<UDP> bind(Endpoint ep) async {
    var udp = UDP._(ep);

    if (ep.isMulticast) {
      var anyAddress = InternetAddress.anyIPv4;
      udp._socket = await RawDatagramSocket.bind(anyAddress, ep.port!.value);
      udp._socket!.joinMulticast(ep.address!);
    } else {
      udp._socket = await RawDatagramSocket.bind(ep.address!, ep.port!.value);
    }

    return udp;
  }

  /// Sends some [data] to a [remoteEndpoint].
  ///
  /// [data] - the data to send.
  /// [remoteEndpoint] - the remote endpoint.
  ///
  /// returns the number of bytes sent.
  /// if the udp was already closed at the time of the call, returns -1.
  Future<int> send(List<int> data, Endpoint remoteEndpoint) async {
    if (_socket == null || _closed) return -1;

    return Future.microtask(() async {
      var prevState = _socket!.broadcastEnabled;

      if (remoteEndpoint.isBroadcast) {
        _socket!.broadcastEnabled = true;
      }

      var _dataCount = _socket!
          .send(data, remoteEndpoint.address!, remoteEndpoint.port!.value);

      _socket!.broadcastEnabled = prevState;

      return _dataCount;
    });
  }

  /// Tells the [UDP] instance to listen for incoming messages.
  ///
  /// Optionally, a [timeout] can be specified. if it is, the [UDP] instance
  /// stops listening after the duration has passed. The instance is closed too,
  /// making the instance unusable after the timeout runs out.
  ///
  /// Whenever new data is received, it is bundled in a [Datagram] and pushed into the stream.
  ///
  /// Returns a [Stream] that can be listened to.
  Stream<Datagram?> asStream({Duration? timeout}) {
    _streamController ??= StreamController<Datagram>();

    _udpBroadcastStream ??= (_streamController as StreamController<Datagram?>)
        .stream
        .asBroadcastStream();

    if (_socket == null || _closed) return _udpBroadcastStream!;

    if (_socketBroadcastStream == null) {
      _socketBroadcastStream = _socket!.asBroadcastStream();

      var streamSubscription = _socketBroadcastStream!.listen((event) {
        if (event == RawSocketEvent.read) {
          (_streamController as StreamController<Datagram?>)
              .add(_socket!.receive());
        }
      });

      if (!_streamSubscriptions.contains(streamSubscription)) {
        _streamSubscriptions.add(streamSubscription);
      }
    }

    if (timeout == null) return _udpBroadcastStream!;

    Future.delayed(timeout).then((value) => close());

    return _udpBroadcastStream!;
  }

  /// Closes the [UDP] instance and the underlying socket.
  void close() {
    _closed = true;
    _socket?.close();
    _socket = null;
    _socketBroadcastStream = null;
    _streamController?.close();
    _streamSubscriptions.forEach((streamSubscription) {
      streamSubscription.cancel();
    });
    _streamSubscriptions.clear();
  }
}


lib/src/udp_port.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright
 *  notice, this list of conditions and the following disclaimer in the
 *  documentation and/or other materials provided with the distribution.
 *
 *  3. Neither the name of the copyright holder nor the names of its
 *  contributors may be used to endorse or promote products derived from
 *  this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 */

import 'dart:io';

import 'package:udp/src/udp_endpoint.dart';

/// [Port] represents an endpoint port.
///
/// Together with [InternetAddress] it defines an [Endpoint].
/// [Endpoint]s are destinations for packets.
///
///
/// See Also:
/// [Endpoint]
class Port {
  /// the value of the Port
  final int value;

  /// [any] represents a random port available on the machine.
  ///
  /// It is up to the OS to pick an appropriate port.
  static const any = Port(0);

  /// Creates a new [Port]
  const Port(this.value);
}


lib/src/udp_endpoint.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *  2. Redistributions in binary form must reproduce the above copyright
 *  notice, this list of conditions and the following disclaimer in the
 *  documentation and/or other materials provided with the distribution.
 *
 *  3. Neither the name of the copyright holder nor the names of its
 *  contributors may be used to endorse or promote products derived from
 *  this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 */

import 'dart:io';

import 'udp_port.dart';

/// [Endpoint] represents a destination for UDP packets.
///
/// Bundles an [InternetAddress] and a [Port].
class Endpoint {
  /// The address of this endpoint.
  InternetAddress? _address;

  /// The address of this endpoint.
  InternetAddress? get address => _address;

  /// The port of this endpoint.
  Port? _port;

  /// The port of this endpoint.
  Port? get port => _port;

  /// Whether the endpoint is a broadcast endpoint.
  bool _isBroadcast = false;

  /// Whether the endpoint is a broadcast endpoint.
  bool get isBroadcast => _isBroadcast;

  /// Whether the endpoint is a multicast endpoint.
  bool _isMulticast = false;

  /// Whether the endpoint is a multicast endpoint.
  bool get isMulticast => _isMulticast;

  /// Creates a Unicast endpoint.
  ///
  /// [_address] is the address to use for binding or sending.
  /// 
  /// [port] represents the port the endpoint is bound to.
  /// 
  /// A UDP instance can bind to a [Endpoint.unicast] endpoint only if any of the network interfaces available is bound to the [Endpoint.unicast] endpoint's address.
  /// As a sender, a UDP instance can send data to [Endpoint.unicast] endpoints.
  Endpoint.unicast(this._address, {Port port = Port.any}) {
    _port = port;
  }

  /// Creates a Broadcast endpoint.
  ///
  /// [port] represents the port the endpoint is bound to.
  /// 
  /// A UDP instance cannot bind to a [Endpoint.broadcast] endpoint.
  /// As a sender, a UDP instance can send data to the [Endpoint.broadcast] endpoint.
  Endpoint.broadcast({Port port = Port.any}) {
    _address = InternetAddress('255.255.255.255');
    _port = port;
    _isBroadcast = true;
  }

  /// Creates a Loopback endpoint bound to the address of the local machine (127.0.0.1).
  ///
  /// [port] represents the port the endpoint is bound to.
  /// 
  /// A UDP instance can bind to a [Endpoint.loopback] endpoint as a sender or as a receiver.
  /// As a sender, a UDP instance can send data to a [Endpoint.loopback] endpoint.
  Endpoint.loopback({Port port = Port.any}) {
    _address = InternetAddress.loopbackIPv4;
    _port = port;
  }

  /// Creates a Multicast endpoint.
  ///
  /// [_address] should be a valid Multicast address in the range 224.0.0.0
  /// to 239.255.255.255.
  /// 
  /// [port] represents the port the endpoint is bound to.
  /// 
  /// A UDP instance can bind to a [Endpoint.multicast] endpoint only as a receiver.
  /// As a sender, a UDP instance can send data to [Endpoint.multicast] endpoints.
  Endpoint.multicast(this._address, {Port port = Port.any}) {
    _port = port;
    _isMulticast = true;
  }

  /// Creates a random Endpoint.
  ///
  /// [port] represents the port the endpoint is bound to.
  /// 
  /// The OS will choose an appropriate [InternetAddress] and [Port].
  /// A UDP instance can bind to a [Endpoint.any] endpoint as a sender or as a receiver.
  /// As a sender, a UDP instance can send data to [Endpoint.any] endpoints.
  Endpoint.any({port = Port.any}) {
    _address = InternetAddress.anyIPv4;
    _port = port;
  }

  /// internal constructor.
  // ignore_for_file: unused_element
  Endpoint._(this._address, this._port);
}


example/udp_example_multicast.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha 
 *  
 *  Redistribution and use in source and binary forms, with or without 
 *  modification, are permitted provided that the following conditions 
 *  are met: 
 *  
 *  1. Redistributions of source code must retain the above copyright 
 *  notice, this list of conditions and the following disclaimer. 
 *  
 *  2. Redistributions in binary form must reproduce the above copyright 
 *  notice, this list of conditions and the following disclaimer in the 
 *  documentation and/or other materials provided with the distribution. 
 *  
 *  3. Neither the name of the copyright holder nor the names of its 
 *  contributors may be used to endorse or promote products derived from 
 *  this software without specific prior written permission. 
 *  
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY 
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 *  POSSIBILITY OF SUCH DAMAGE.
 *  
 */

import 'dart:io';

import 'package:udp/udp.dart';

void main() async {
  // MULTICAST
  var multicastEndpoint =
      Endpoint.multicast(InternetAddress('239.1.2.3'), port: Port(54321));

  var receiver = await UDP.bind(multicastEndpoint);

  var sender = await UDP.bind(Endpoint.any());

  receiver.asStream().listen((datagram) {
    if (datagram != null) {
      var str = String.fromCharCodes(datagram.data);

      stdout.write(str);
    }
  });

  await sender.send('Foo'.codeUnits, multicastEndpoint);

  await Future.delayed(Duration(seconds: 5));

  sender.close();
  receiver.close();
}


example/udp_example.dart
/*
 *
 *  Copyright 2019-2022 Kennedy Tochukwu Ekeoha 
 *  
 *  Redistribution and use in source and binary forms, with or without 
 *  modification, are permitted provided that the following conditions 
 *  are met: 
 *  
 *  1. Redistributions of source code must retain the above copyright 
 *  notice, this list of conditions and the following disclaimer. 
 *  
 *  2. Redistributions in binary form must reproduce the above copyright 
 *  notice, this list of conditions and the following disclaimer in the 
 *  documentation and/or other materials provided with the distribution. 
 *  
 *  3. Neither the name of the copyright holder nor the names of its 
 *  contributors may be used to endorse or promote products derived from 
 *  this software without specific prior written permission. 
 *  
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
 *  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
 *  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED 
 *  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY 
 *  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
 *  POSSIBILITY OF SUCH DAMAGE.
 *  
 */

import 'dart:io';

import 'package:udp/udp.dart';

void main() async {
  // creates a UDP instance and binds it to the first available network
  // interface on port 65000.
  var sender = await UDP.bind(Endpoint.any(port: Port(65000)));

  // send a simple string to a broadcast endpoint on port 65001.
  var dataLength = await sender.send(
      'Hello World!'.codeUnits, Endpoint.broadcast(port: Port(65001)));

  stdout.write('$dataLength bytes sent.');

  // creates a new UDP instance and binds it to the local address and the port
  // 65002.
  var receiver = await UDP.bind(Endpoint.loopback(port: Port(65002)));

  // receiving\listening
  receiver.asStream(timeout: Duration(seconds: 20)).listen((datagram) {
    var str = String.fromCharCodes(datagram!.data);
    stdout.write(str);
  });

  // close the UDP instances and their sockets.
  sender.close();
  receiver.close();
}
